### ISSUE TITLE
Feature provider with id 'service' could not be started: Failed to initialize pool: FATAL: password authentication failed for user "service" 

### ISSUE THREAD

**Issue by swiss-knight** [interactive-instruments/ldproxy#583]
I'm trying to follow the tutorial, using a self hosted database but I cannot make it work, here's the log:

```
$ cd docker/ldproxy_data

$ tail -f log/xtraplatform.log

INFO  [2022-01-25 22:25:04,598]                          - Starting ldproxy 3.1.0  
INFO  [2022-01-25 22:25:10,142]                          - Store mode: READ_WRITE  
INFO  [2022-01-25 22:25:10,142]                          - Store location: /ldproxy/data/store  
INFO  [2022-01-25 22:25:17,321]                          - Started web server at http://localhost:7080  
ERROR [2022-01-25 22:25:21,879] service                 - Feature provider with id 'service' could not be started: Failed to initialize pool: FATAL: password authentication failed for user "service"  
ERROR [2022-01-25 22:25:22,347] service                 - Entity of type 'services/ogc_api' with id 'service' could not be hydrated: No feature provider found.  
INFO  [2022-01-25 22:25:22,519] service                 - Service with id 'service' started successfully.  
ERROR [2022-01-25 22:25:23,058] service                 - Feature provider with id 'service' could not be started: Failed to initialize pool: FATAL: password authentication failed for user "service" 
```

I'm running it the way described in the [documentation](https://github.com/interactive-instruments/ldproxy/blob/master/docs/en/deployment.md):
```
$ docker run \ 
  --name ldproxy \
  --add-host=host.docker.internal:host-gateway \
  -d \
  -p 7080:7080 \
  -v ~/docker/ldproxy_data:/ldproxy/data \
  iide/ldproxy:latest
```

Here's the config: 

```
$ cat cfg.yml 
---
server:
  applicationConnectors:
    - type: http
      port: '7080'
  adminConnectors:
    - type: http
      port: '7081'
  gzip:
    enabled: true
  requestLog:
    type: classic
    timeZone: Europe/Berlin
    appenders:
      - type: file
        currentLogFilename: "./data/log/requests.log"
        archive: true
        archivedLogFilenamePattern: "./data/log/requests-%d.log.gz"
        archivedFileCount: 7
  shutdownGracePeriod: 30s
httpClient:
  timeout: 30000ms
  connectionTimeout: 30000ms
logging:
  level: 'OFF'
  appenders:
    - type: file
      timeZone: Europe/Berlin
      currentLogFilename: "./data/log/xtraplatform.log"
      archive: true
      archivedLogFilenamePattern: "./data/log/xtraplatform-%d.log.gz"
      archivedFileCount: 7
      logFormat: "%-5p [%d{ISO8601,UTC}] %-48.48c{40}: %m%n%rEx"
  loggers:
    de.ii: INFO 
```

and:
```
$ cat store/entities/providers/service.yml
---
id: service
createdAt: 1643145562759
lastModified: 1643145562759
entityStorageVersion: 2
providerType: FEATURE
featureProviderType: SQL
connectionInfo:
  database: service
  host: host.docker.internal
  user: service
  port: 5433
  password: <base64_password>
```

the password was encoded using `echo -n 'password_string' | base64` with:    
```
$  base64 --version
base64 (GNU coreutils) 8.28
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Simon Josefsson.
```

and:
```
$ cat store/entities/services/service.yml 
---
id: service
createdAt: 1643145562848
lastModified: 1643145562848
entityStorageVersion: 2
label: service
serviceType: OGC_API
```

I'm also facing this error when browsing this URL with Firefox (96.0.2 (64-bit)):    
`http://my-base-server:7080/rest/services/service/collections`
```
{
  "code" : 500,
  "message" : "There was an error processing your request. It has been logged (ID f7186328fa162ee8)."
}
```

OS: Ubuntu 18.04.3

ps1: `docker logs -f ldproxy` shows absolutely nothing, even when I'm browsing or getting some endpoint using `curl`.    
ps2: I cannot find any documentation on how to specify a different port from the default postgresql one `5432`.

Btw, thanks for your great job and wonderful demos out [there](https://demo.ldproxy.net/)!

**Comment by cportele** [comment]
@swiss-knight - Thanks for trying out ldproxy. Some answers and comments:

### DB connection failure

I suspect that the DB connection issue could be a result of

> ps2: I cannot find any documentation on how to specify a different port from the default postgresql one 5432.

and

```yaml
connectionInfo:
  database: service
  host: host.docker.internal
  user: service
  port: 5433
  password: <base64_password>
```

If you do not use 5432 as the port, you have to add it as part of the "host" (not via another member). In your case: `host: host.docker.internal:5433`.

This is described in the [SQL provider configuration doc](https://github.com/interactive-instruments/ldproxy/blob/master/docs/en/configuration/providers/sql.md):

> To use a non-default port, add it to the host separated by `:`, e.g. `db:30305`.

NB: The ldproxy documentation is not complete at the moment (at least the one in English, the German documentation should cover the current capabilities), it is a task for this year to revise the documentation and make it easier to manage.

Since you are on linux, I don't think you need to use 'host.docker.internal', this is only required for Docker Desktop on macOS or Windows AFAIK (I am using a Mac). You should be able to use the standard [docker network settings](https://docs.docker.com/engine/reference/run/#network-settings). Depending on the mode you choose you would connect to your database running on the docker host in different ways.

Could it be that you also have an instance at the default port, with a user "service" and a different password? The error message "FATAL: password authentication failed for user 'service'" indicates that the DB connection could be established, that there is a user "service", but that the password is incorrect.

If that is not the case, please double check that you have used the correct password, `echo -n 'password_string' | base64` looks correct.

### `docker logs -f ldproxy` empty

> ps1: docker logs -f ldproxy shows absolutely nothing, even when I'm browsing or getting some endpoint using curl.

The reason are the settings in your global configuration `cfg.yml`. If there is no `cfg.yml` file at startup, ldproxy creates one that looks like the one you have. I think we should update the behaviour to create a minimal default configuration.

The configuration created by ldproxy writes the log to the `log` folder in the data directory. Youe log is in `log/xtraplatform.log`. 

To get the log via `docker logs` change the output from the file to the console. That is, replace the `appenders` key

```yaml
logging:
  appenders:
    - type: file
      timeZone: Europe/Berlin
      currentLogFilename: "./data/log/xtraplatform.log"
      archive: true
      archivedLogFilenamePattern: "./data/log/xtraplatform-%d.log.gz"
      archivedFileCount: 7
      logFormat: "%-5p [%d{ISO8601,UTC}] %-48.48c{40}: %m%n%rEx"
```

with

```yaml
logging:
  appenders:
    - type: console
```

Note that to see information about requests in the standard log, you will have to change the log level to DEBUG, i.e., replace `de.ii: INFO` with `de.ii: DEBUG`.

Request logging is switched off by default, but it can be activated in `cfg.yml`, as described in the [documentation](https://github.com/interactive-instruments/ldproxy/blob/master/docs/de/configuration/global-configuration.md#request-logging) (currently only in German, sorry). This is included in your `cfg.yml`, so you should see the requests in the request log.

### 500 errors

The "service" API should not start, because the feature provider could not be created. For some reason the checks do seem to be complete and I will create another issue for this. I suspect that the 500 errors that you get, when you access `/service/collections` are very likely a result of this.

The reason that no information about the error is included in the log file is a known issue (#569). If you switch the log level to DEBUG, you will see something like `Sending response: 500 Internal Server Error`, which of course is no help as it has no explanation for the cause of the error. We hope to fix this in the next release.

### provider and service configurations

Your current provider and service configurations will not create a useful API, because no feature collections will be available. To derive the information about feature types from the database, add `auto: true` to both configurations. Then ldproxy will create a feature type and collection for every table in the "public" schema of the "service" database during startup and try to determine some other information (like the CRS of the geometries in the DB). 

To run the analysis once only, add also `autoPersist: true` in both configuration files. Then the information about the feature types and collections will be written to the configuration files (and you can adapt it). 

Something like:

```bash
$ cat store/entities/providers/service.yml
---
id: service
entityStorageVersion: 2
providerType: FEATURE
featureProviderType: SQL
connectionInfo:
  database: service
  host: host.docker.internal:5433
  user: service
  password: <base64_password>
auto: true
autoPersist: true

$ cat store/entities/services/service.yml 
---
id: service
entityStorageVersion: 2
label: service
serviceType: OGC_API
auto: true
autoPersist: true
```

**Comment by swiss-knight** [comment]
Many thanks for your really rich answer.

Ok, I append the port to the hostname, as suggested [here](https://github.com/interactive-instruments/ldproxy/blob/master/docs/en/configuration/providers/sql.md), and I also added the `auto: true` and `autoPersist: true`, to both the provider and the service yaml files.     
These three changes helped a lot, and I am now successful reaching: http://my-base-server:7080/rest/services/service/collections!    
The log doesn't show the error anymore, and with the `DEBUG` level, it also shows, e.g. the seeding of the different collections from the database. That's really nice!

But I'm facing a new error #587
